src/hotspot/share/opto/loopopts.cpp

```diff
@@ -190,7 +190,7 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
 
     if (x->is_Con()) {
       // Constant's control is always root.
-      set_ctrl(x, C->root());
+      set_root_as_ctrl(x);
       continue;
     }
```

Comment:
This looks like "band-aid" - this should be assert. May be investigate in separate RFE.

Reply:
I opened an RFE for this https://bugs.openjdk.org/browse/JDK-8343907

Reply:
If you modify the following code above to use your new `makecon()` (could be done either way), could this then be turned into an assert? By looking at the code, it suggests that we only miss to set ctrl in the `singleton` case which would then be covered.
https://github.com/openjdk/jdk/blob/5ca6698ba418e82ff93471fbb495759850f26f63/src/hotspot/share/opto/loopopts.cpp#L123-L125
You could also only change `makecon()` above and revisit this code later again to remove the `set_root_as_ctrl()` and add an assert.

Reply:
I added the asserts as @chhagedorn suggested. All tests in the internal testing passed with this new assert.